solv/tdnfpackage.c

```diff
@@ -1924,3 +1924,302 @@ SolvGetNevraFromId(
+            queue_free(&queueTmp);
+        }
+    }
+    else if (idKey == pool_str2id(pSack->pPool, TDNF_ID_REQUIRES_PRE, 0))
+    {
+        solvable_lookup_deparray(pSolv, SOLVABLE_REQUIRES, &queueDeps, 1);
+    }
+    else
+    {
+        solvable_lookup_deparray(pSolv, idKey, &queueDeps, -1);
+    }
+    nNumDeps = queueDeps.count;
+
+    dwError = TDNFAllocateMemory(nNumDeps + 1, sizeof(char *), (void**)&ppszDependencies);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    for (i = 0; i < nNumDeps; i++)
+    {
+        pszDep = pool_id2str(pSack->pPool, queueDeps.elements[i]);
+        dwError = TDNFAllocateString(pszDep, &ppszDependencies[i]);
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    *pppszDependencies = ppszDependencies;
+
+cleanup:
+    queue_free(&queueDeps);
+    return dwError;
+error:
+    TDNFFreeStringArray(ppszDependencies);
+    goto cleanup;
+}
+
+uint32_t
+SolvGetFileListFromId(
+    PSolvSack pSack,
+    uint32_t dwPkgId,
+    char ***pppszFiles)
+{
+    uint32_t dwError = 0;
+    Solvable *pSolv = NULL;
+    Dataiterator di;
+    int i = 0, nCount = 0;
+    char **ppszFiles = NULL;
+
+    if(!pSack || !pppszFiles)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_LIBSOLV_ERROR(dwError);
+    }
+
+    pSolv = pool_id2solvable(pSack->pPool, dwPkgId);
+    if(!pSolv)
+    {
+        dwError = ERROR_TDNF_NO_DATA;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    dataiterator_init(&di, pSack->pPool, pSolv->repo, dwPkgId, SOLVABLE_FILELIST, NULL,
+                      SEARCH_FILES | SEARCH_COMPLETE_FILELIST);
+    while (dataiterator_step(&di)) {
+        nCount++;
+    }
+    dataiterator_free(&di);
+
+    dwError = TDNFAllocateMemory(nCount + 1, sizeof(char *), (void**)&ppszFiles);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    dataiterator_init(&di, pSack->pPool, pSolv->repo, dwPkgId, SOLVABLE_FILELIST, NULL,
+                      SEARCH_FILES | SEARCH_COMPLETE_FILELIST);
+    while (dataiterator_step(&di)) {
+        dwError = TDNFAllocateString(di.kv.str, &ppszFiles[i++]);
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+    dataiterator_free(&di);
+
+    *pppszFiles = ppszFiles;
+cleanup:
+    return dwError;
+error:
+    TDNFFreeStringArray(ppszFiles);
+    goto cleanup;
+}
+
+uint32_t
+SolvGetSourceFromId(
+    PSolvSack pSack,
+    uint32_t dwPkgId,
+    char **ppszName,
+    char **ppszArch,
+    char **ppszEVR
+    )
+{
+    uint32_t dwError = 0;
+    const char* pszTmp = NULL;
+    Solvable *pSolv = NULL;
+    char *pszName = NULL;
+    char *pszArch = NULL;
+    char *pszEVR = NULL;
+
+    if(!pSack ||
+       !ppszName ||
+       !ppszArch ||
+       !ppszEVR)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_LIBSOLV_ERROR(dwError);
+    }
+
+    pSolv = pool_id2solvable(pSack->pPool, dwPkgId);
+    if(!pSolv)
+    {
+        dwError = ERROR_TDNF_NO_DATA;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    pszTmp = solvable_lookup_str(pSolv, SOLVABLE_SOURCENAME);
```

Comment:
Unfortunately `SOLVABLE_SOURCENAME`/`SOLVABLE_SOURCEEVR` and `SOLVABLE_SOURCEARCH` are handled differently so we can only loop over the first two. I don't think it would make nicer.